File: windows/AudioTranscriptionSummary/Services/SSOAuthService.cs
// SSOAuthService.cs
// IAM Identity Center（SSO）の OIDC Device Authorization Flow を管理するサービス
// RegisterClient → StartDeviceAuthorization → CreateToken ポーリング → アカウント/ロール選択 → 一時認証情報取得

#nullable enable
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using Amazon.SSO;
using Amazon.SSO.Model;
using Amazon.SSOOIDC;
using Amazon.SSOOIDC.Model;
using AudioTranscriptionSummary.Models;

namespace AudioTranscriptionSummary.Services;

/// <summary>
/// SSO 認証フロー全体を管理するサービスクラス（シングルトン）
/// OIDC Device Authorization Grant（RFC 8628）を使用してブラウザ経由でユーザー認証を行う
/// </summary>
public class SSOAuthService : INotifyPropertyChanged
{
    // シングルトンインスタンス
    public static SSOAuthService Instance { get; } = new SSOAuthService();

    private SSOAuthService() { }

    // MARK: - プロパティ

    private SSOLoginState _loginState = SSOLoginState.Idle;
    /// <summary>現在のログイン状態</summary>
    public SSOLoginState LoginState
    {
        get => _loginState;
        set => SetField(ref _loginState, value);
    }

    private string? _errorMessage;
    /// <summary>エラーメッセージ</summary>
    public string? ErrorMessage
    {
        get => _errorMessage;
        set => SetField(ref _errorMessage, value);
    }

    private string? _userCode;
    /// <summary>ブラウザで入力する User Code</summary>
    public string? UserCode
    {
        get => _userCode;
        set => SetField(ref _userCode, value);
    }

    private string? _verificationUri;
    /// <summary>認証 URI</summary>
    public string? VerificationUri
    {
        get => _verificationUri;
        set => SetField(ref _verificationUri, value);
    }

    private List<SSOAccountInfo> _accounts = new();
    /// <summary>利用可能なアカウント一覧</summary>
    public List<SSOAccountInfo> Accounts
    {
        get => _accounts;
        set => SetField(ref _accounts, value);
    }

    private List<string> _roles = new();
    /// <summary>選択されたアカウントで利用可能なロール一覧</summary>
    public List<string> Roles
    {
        get => _roles;
        set => SetField(ref _roles, value);
    }

    private SSOTemporaryCredentials? _temporaryCredentials;
    /// <summary>取得した一時認証情報</summary>
    public SSOTemporaryCredentials? TemporaryCredentials
    {
        get => _temporaryCredentials;
        set => SetField(ref _temporaryCredentials, value);
    }

    // MARK: - 内部状態（メモリ内のみ保持）

    /// <summary>SSO Access Token</summary>
    private string? _accessToken;
    /// <summary>Access Token の有効期限</summary>
    private DateTime? _accessTokenExpiry;
    /// <summary>SSO リージョン（認証フロー用）</summary>
    private string? _ssoRegion;

    // MARK: - キャッシュファイルパス

    private static readonly string CacheFilePath = System.IO.Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
        "AudioTranscriptionSummary", ".sso_cache.json");

    // MARK: - キャッシュ保存/復元

    /// <summary>一時認証情報をファイルにキャッシュする</summary>
    public void SaveCacheToFile()
    {
        if (TemporaryCredentials == null) return;
        try
        {
            var cache = new SSOCachedSession
            {
                AccessKeyId = TemporaryCredentials.AccessKeyId,
                SecretAccessKey = TemporaryCredentials.SecretAccessKey,
                SessionToken = TemporaryCredentials.SessionToken,
                Expiration = TemporaryCredentials.Expiration,
                AccessToken = _accessToken,
                AccessTokenExpiry = _accessTokenExpiry,
                SsoRegion = _ssoRegion
            };
            var dir = System.IO.Path.GetDirectoryName(CacheFilePath);
            if (!string.IsNullOrEmpty(dir) && !System.IO.Directory.Exists(dir))
                System.IO.Directory.CreateDirectory(dir);
            var json = System.Text.Json.JsonSerializer.Serialize(cache);
            System.IO.File.WriteAllText(CacheFilePath, json);
        }
        catch { /* キャッシュ保存失敗は致命的ではない */ }
    }

    /// <summary>ファイルからキャッシュを復元する</summary>
    public void RestoreFromCache()
    {
        if (!System.IO.File.Exists(CacheFilePath)) return;
        try
        {
            var json = System.IO.File.ReadAllText(CacheFilePath);
            var cache = System.Text.Json.JsonSerializer.Deserialize<SSOCachedSession>(json);
            if (cache == null || cache.Expiration <= DateTime.UtcNow)
            {
                ClearCache();
                return;
            }
            TemporaryCredentials = new SSOTemporaryCredentials
            {
                AccessKeyId = cache.AccessKeyId,
                SecretAccessKey = cache.SecretAccessKey,
                SessionToken = cache.SessionToken,
                Expiration = cache.Expiration
            };
            _accessToken = cache.AccessToken;
            _accessTokenExpiry = cache.AccessTokenExpiry;
            _ssoRegion = cache.SsoRegion;
            LoginState = SSOLoginState.Authenticated;
        }
        catch
        {
            ClearCache();
        }
    }

    /// <summary>キャッシュファイルを削除する</summary>
    private void ClearCache()
    {
        try { System.IO.File.Delete(CacheFilePath); } catch { }
    }

    // MARK: - Access Token 有効期限判定

    /// <summary>Access Token が有効かどうか</summary>
    public bool IsTokenValid
    {
        get
        {
            if (_accessTokenExpiry == null) return false;
            return _accessTokenExpiry.Value > DateTime.UtcNow;
        }
    }

    // MARK: - OIDC Device Authorization Flow

    /// <summary>
    /// SSO ログインを開始する
    /// RegisterClient → StartDeviceAuthorization → ブラウザ起動 → CreateToken ポーリング
    /// </summary>
    /// <param name="startUrl">IAM Identity Center の Start URL</param>
    /// <param name="region">SSO リージョン</param>
    public async Task StartLoginAsync(string startUrl, string region)
    {
        // 状態をリセット
        LoginState = SSOLoginState.Registering;
        ErrorMessage = null;
        UserCode = null;
        VerificationUri = null;
        Accounts = new List<SSOAccountInfo>();
        Roles = new List<string>();
        TemporaryCredentials = null;
        _accessToken = null;
        _accessTokenExpiry = null;
        _ssoRegion = region;

        try
        {
            // OIDC クライアントを作成
            var oidcConfig = new AmazonSSOOIDCConfig
            {
                RegionEndpoint = Amazon.RegionEndpoint.GetBySystemName(region)
            };
            using var oidcClient = new AmazonSSOOIDCClient(
                new Amazon.Runtime.AnonymousAWSCredentials(), oidcConfig);

            // 1. RegisterClient API 呼び出し
            var registerResponse = await oidcClient.RegisterClientAsync(new RegisterClientRequest
            {
                ClientName = "AudioTranscriptionSummary",
                ClientType = "public"
            });

            var clientId = registerResponse.ClientId;
            var clientSecret = registerResponse.ClientSecret;

            // 2. StartDeviceAuthorization API 呼び出し
            var deviceAuthResponse = await oidcClient.StartDeviceAuthorizationAsync(
                new StartDeviceAuthorizationRequest
                {
                    ClientId = clientId,
                    ClientSecret = clientSecret,
                    StartUrl = startUrl
                });

            var deviceCode = deviceAuthResponse.DeviceCode;
            var userCode = deviceAuthResponse.UserCode;
            var verificationUri = deviceAuthResponse.VerificationUri;
            var interval = deviceAuthResponse.Interval > 0 ? deviceAuthResponse.Interval : 5;

            // User Code を表示
            UserCode = userCode;
            VerificationUri = verificationUri;
            LoginState = SSOLoginState.WaitingForBrowser;

            // 3. ブラウザで認証 URI を開く
            try
            {
                Process.Start(new ProcessStartInfo
                {
                    FileName = verificationUri,
                    UseShellExecute = true
                });
            }
            catch
            {
                // ブラウザ起動失敗は無視（ユーザーが手動で開ける）
            }

            // 4. CreateToken API をポーリング
            LoginState = SSOLoginState.Polling;
            var pollInterval = interval;

            while (true)
            {
                await Task.Delay(pollInterval * 1000);

                try
                {
                    var tokenResponse = await oidcClient.CreateTokenAsync(new CreateTokenRequest
                    {
                        ClientId = clientId,
                        ClientSecret = clientSecret,
                        DeviceCode = deviceCode,
                        GrantType = "urn:ietf:params:oauth:grant-type:device_code"
                    });

                    // 認証成功
                    _accessToken = tokenResponse.AccessToken;
                    if (tokenResponse.ExpiresIn > 0)
                    {
                        _accessTokenExpiry = DateTime.UtcNow.AddSeconds(tokenResponse.ExpiresIn);
                    }

                    LoginState = SSOLoginState.SelectingAccount;
                    return;
                }
                catch (AuthorizationPendingException)
                {
                    // authorization_pending: ポーリング継続
                    continue;
                }
                catch (SlowDownException)
                {
                    // slow_down: ポーリング間隔を5秒延長
                    pollInterval += 5;
                    continue;
                }
                catch (ExpiredTokenException)
                {
                    // expired_token: タイムアウト
                    ErrorMessage = "認証がタイムアウトしました。再度 SSO ログインを実行してください";
                    LoginState = SSOLoginState.Error;
                    return;
                }
            }
        }
        catch (Exception ex)
        {
            ErrorMessage = $"SSO 接続に失敗しました: {ex.Message}";
            LoginState = SSOLoginState.Error;
        }
    }

    // MARK: - アカウント一覧取得

    /// <summary>SSO で利用可能なアカウント一覧を取得する</summary>
    public async Task FetchAccountsAsync()
    {
        if (_accessToken == null || !IsTokenValid)
        {
            ErrorMessage = "SSO セッションが期限切れです。再度 SSO ログインを実行してください";
            LoginState = SSOLoginState.Error;
            return;
        }
        if (_ssoRegion == null)
        {
            ErrorMessage = "SSO リージョンが設定されていません";
            LoginState = SSOLoginState.Error;
            return;
        }

        try
        {
            var ssoConfig = new AmazonSSOConfig
            {
                RegionEndpoint = Amazon.RegionEndpoint.GetBySystemName(_ssoRegion)
            };
            using var ssoClient = new AmazonSSOClient(
                new Amazon.Runtime.AnonymousAWSCredentials(), ssoConfig);

            var response = await ssoClient.ListAccountsAsync(new ListAccountsRequest
            {
                AccessToken = _accessToken
            });

            var accountList = response.AccountList ?? new List<AccountInfo>();
            Accounts = accountList
                .Where(a => !string.IsNullOrEmpty(a.AccountId) && !string.IsNullOrEmpty(a.AccountName))
                .Select(a => new SSOAccountInfo
                {
                    AccountId = a.AccountId,
                    AccountName = a.AccountName
                })
                .ToList();

            if (Accounts.Count == 0)
            {
                ErrorMessage = "利用可能なアカウントがありません。IAM Identity Center の権限設定を確認してください";
                LoginState = SSOLoginState.Error;
            }
            else
            {
                LoginState = SSOLoginState.SelectingAccount;
            }
        }
        catch (Exception ex)
        {
            ErrorMessage = $"アカウント一覧の取得に失敗しました: {ex.Message}";
            LoginState = SSOLoginState.Error;
        }
    }

    // MARK: - ロール一覧取得

    /// <summary>指定アカウントで利用可能なロール一覧を取得する</summary>
    /// <param name="accountId">AWS アカウント ID</param>
    public async Task FetchRolesAsync(string accountId)
    {
        if (_accessToken == null || !IsTokenValid)
        {
            ErrorMessage = "SSO セッションが期限切れです。再度 SSO ログインを実行してください";
            LoginState = SSOLoginState.Error;
            return;
        }
        if (_ssoRegion == null)
        {
            ErrorMessage = "SSO リージョンが設定されていません";
            LoginState = SSOLoginState.Error;
            return;
        }

        try
        {
            var ssoConfig = new AmazonSSOConfig
            {
                RegionEndpoint = Amazon.RegionEndpoint.GetBySystemName(_ssoRegion)
            };
            using var ssoClient = new AmazonSSOClient(
                new Amazon.Runtime.AnonymousAWSCredentials(), ssoConfig);

            var response = await ssoClient.ListAccountRolesAsync(new ListAccountRolesRequest
            {
                AccessToken = _accessToken,
                AccountId = accountId
            });

            var roleList = response.RoleList ?? new List<RoleInfo>();
            Roles = roleList
                .Where(r => !string.IsNullOrEmpty(r.RoleName))
                .Select(r => r.RoleName)
                .ToList();

            if (Roles.Count == 0)
            {
                ErrorMessage = "このアカウントで利用可能なロールがありません";
                LoginState = SSOLoginState.Error;
            }
            else
            {
                LoginState = SSOLoginState.SelectingRole;
            }
        }
        catch (Exception ex)
        {
            ErrorMessage = $"ロール一覧の取得に失敗しました: {ex.Message}";
            LoginState = SSOLoginState.Error;
        }
    }

    // MARK: - 一時認証情報取得

    /// <summary>指定アカウント・ロールの一時認証情報を取得する</summary>
    /// <param name="accountId">AWS アカウント ID</param>
    /// <param name="roleName">ロール名</param>
    public async Task FetchCredentialsAsync(string accountId, string roleName)
    {
        if (_accessToken == null || !IsTokenValid)
        {
            ErrorMessage = "SSO セッションが期限切れです。再度 SSO ログインを実行してください";
            LoginState = SSOLoginState.Error;
            return;
        }
        if (_ssoRegion == null)
        {
            ErrorMessage = "SSO リージョンが設定されていません";
            LoginState = SSOLoginState.Error;
            return;
        }

        try
        {
            var ssoConfig = new AmazonSSOConfig
            {
                RegionEndpoint = Amazon.RegionEndpoint.GetBySystemName(_ssoRegion)
            };
            using var ssoClient = new AmazonSSOClient(
                new Amazon.Runtime.AnonymousAWSCredentials(), ssoConfig);

            var response = await ssoClient.GetRoleCredentialsAsync(new GetRoleCredentialsRequest
            {
                AccessToken = _accessToken,
                AccountId = accountId,
                RoleName = roleName
            });

            var creds = response.RoleCredentials;
            if (creds == null ||
                string.IsNullOrEmpty(creds.AccessKeyId) ||
                string.IsNullOrEmpty(creds.SecretAccessKey) ||
                string.IsNullOrEmpty(creds.SessionToken))
            {
                ErrorMessage = "一時認証情報の取得に失敗しました";
                LoginState = SSOLoginState.Error;
                return;
            }

            // Expiration はミリ秒単位の Unix タイムスタンプ
            var expiration = creds.Expiration != 0
                ? DateTimeOffset.FromUnixTimeMilliseconds(creds.Expiration).UtcDateTime
                : DateTime.UtcNow.AddHours(1); // デフォルト: 1時間後

            TemporaryCredentials = new SSOTemporaryCredentials
            {
                AccessKeyId = creds.AccessKeyId,
                SecretAccessKey = creds.SecretAccessKey,
                SessionToken = creds.SessionToken,
                Expiration = expiration
            };

            LoginState = SSOLoginState.Authenticated;
            SaveCacheToFile();
        }
        catch (Exception ex)
        {
            ErrorMessage = $"一時認証情報の取得に失敗しました: {ex.Message}";
            LoginState = SSOLoginState.Error;
        }
    }

    // MARK: - リセット

    /// <summary>ログイン状態を完全にリセットする</summary>
    public void Reset()
    {
        LoginState = SSOLoginState.Idle;
        ErrorMessage = null;
        UserCode = null;
        VerificationUri = null;
        Accounts = new List<SSOAccountInfo>();
        Roles = new List<string>();
        TemporaryCredentials = null;
        _accessToken = null;
        _accessTokenExpiry = null;
        _ssoRegion = null;
        ClearCache();
    }

    // MARK: - INotifyPropertyChanged

    public event PropertyChangedEventHandler? PropertyChanged;

    private void OnPropertyChanged([CallerMemberName] string? propertyName = null)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }

    private bool SetField<T>(ref T field, T value, [CallerMemberName] string? propertyName = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return false;
        field = value;
        OnPropertyChanged(propertyName);
        return true;
    }
}

File: macos/Sources/AudioTranscriptionSummary/Services/SSOAuthService.swift
// SSOAuthService.swift
// IAM Identity Center（SSO）の OIDC Device Authorization Flow を管理するサービス
// RegisterClient → StartDeviceAuthorization → CreateToken ポーリング → アカウント/ロール選択 → 一時認証情報取得

import Foundation
import AWSSSOOIDC
import AWSSSO
import AppKit

// MARK: - SSOAuthService

/// SSO 認証フロー全体を管理するサービスクラス
/// OIDC Device Authorization Grant（RFC 8628）を使用してブラウザ経由でユーザー認証を行う
@MainActor
class SSOAuthService: ObservableObject {

    // MARK: - シングルトン

    /// 共有インスタンス
    static let shared = SSOAuthService()

    /// AWSClientFactory から同期的にアクセスするための一時認証情報キャッシュ
    /// @MainActor の temporaryCredentials が更新されるたびに同期される
    nonisolated(unsafe) static var cachedCredentials: SSOTemporaryCredentials?

    // MARK: - Published プロパティ

    /// 現在のログイン状態
    @Published var loginState: SSOLoginState = .idle
    /// 利用可能なアカウント一覧
    @Published var accounts: [SSOAccountInfo] = []
    /// 選択されたアカウントで利用可能なロール一覧
    @Published var roles: [String] = []
    /// 取得した一時認証情報
    @Published var temporaryCredentials: SSOTemporaryCredentials? {
        didSet {
            // AWSClientFactory から同期アクセスできるようキャッシュを更新
            SSOAuthService.cachedCredentials = temporaryCredentials
        }
    }

    // MARK: - 内部状態（メモリ内のみ保持）

    /// SSO Access Token
    private var accessToken: String?
    /// Access Token の有効期限
    private var accessTokenExpiry: Date?
    /// SSO リージョン（認証フロー用）
    private var ssoRegion: String?

    // MARK: - キャッシュファイルパス

    /// 一時認証情報のキャッシュファイルパス
    private static let cacheFilePath: URL = {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            .appendingPathComponent("AudioTranscriptionSummary")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent(".sso_cache.json")
    }()

    // MARK: - キャッシュ保存/復元

    /// 一時認証情報をファイルにキャッシュする
    private func saveCacheToFile() {
        guard let creds = temporaryCredentials else { return }
        let cache = SSOCachedSession(
            credentials: creds,
            accessToken: accessToken,
            accessTokenExpiry: accessTokenExpiry,
            ssoRegion: ssoRegion
        )
        do {
            let data = try JSONEncoder().encode(cache)
            try data.write(to: Self.cacheFilePath, options: .atomic)
        } catch {
            // キャッシュ保存失敗は致命的ではない
        }
    }

    /// ファイルからキャッシュを復元する
    func restoreFromCache() {
        guard FileManager.default.fileExists(atPath: Self.cacheFilePath.path) else { return }
        do {
            let data = try Data(contentsOf: Self.cacheFilePath)
            let cache = try JSONDecoder().decode(SSOCachedSession.self, from: data)

            // 有効期限チェック
            guard cache.credentials.expiration > Date() else {
                // 期限切れ: キャッシュを削除
                try? FileManager.default.removeItem(at: Self.cacheFilePath)
                return
            }

            // 復元
            temporaryCredentials = cache.credentials
            accessToken = cache.accessToken
            accessTokenExpiry = cache.accessTokenExpiry
            ssoRegion = cache.ssoRegion
            loginState = .authenticated
        } catch {
            try? FileManager.default.removeItem(at: Self.cacheFilePath)
        }
    }

    /// キャッシュファイルを削除する
    private func clearCache() {
        try? FileManager.default.removeItem(at: Self.cacheFilePath)
    }

    // MARK: - Access Token 有効期限判定

    /// Access Token が有効かどうか
    var isTokenValid: Bool {
        guard let expiry = accessTokenExpiry else { return false }
        return expiry > Date()
    }

    // MARK: - OIDC Device Authorization Flow

    /// SSO ログインを開始する
    /// RegisterClient → StartDeviceAuthorization → ブラウザ起動 → CreateToken ポーリング
    /// - Parameters:
    ///   - startUrl: IAM Identity Center の Start URL
    ///   - region: SSO リージョン
    func startLogin(startUrl: String, region: String) async throws {
        // 状態をリセット
        loginState = .registering
        accounts = []
        roles = []
        temporaryCredentials = nil
        accessToken = nil
        accessTokenExpiry = nil
        ssoRegion = region

        do {
            // OIDC クライアントを作成
            let oidcConfig = try await SSOOIDCClient.SSOOIDCClientConfiguration(region: region)
            let oidcClient = SSOOIDCClient(config: oidcConfig)

            // 1. RegisterClient API 呼び出し
            let registerInput = RegisterClientInput(
                clientName: "AudioTranscriptionSummary",
                clientType: "public"
            )
            let registerOutput = try await oidcClient.registerClient(input: registerInput)

            guard let clientId = registerOutput.clientId,
                  let clientSecret = registerOutput.clientSecret else {
                throw SSOAuthError.registerClientFailed
            }

            // 2. StartDeviceAuthorization API 呼び出し
            let deviceAuthInput = StartDeviceAuthorizationInput(
                clientId: clientId,
                clientSecret: clientSecret,
                startUrl: startUrl
            )
            let deviceAuthOutput = try await oidcClient.startDeviceAuthorization(input: deviceAuthInput)

            guard let deviceCode = deviceAuthOutput.deviceCode,
                  let userCode = deviceAuthOutput.userCode,
                  let verificationUri = deviceAuthOutput.verificationUri else {
                throw SSOAuthError.startDeviceAuthorizationFailed
            }

            let interval = deviceAuthOutput.interval ?? 5

            // 3. ブラウザで認証 URI を開く
            if let url = URL(string: verificationUri) {
                NSWorkspace.shared.open(url)
            }

            // User Code を表示（ポーリング中も維持）
            loginState = .waitingForBrowser(userCode: userCode, verificationUri: verificationUri)

            // 4. CreateToken API をポーリング（loginState は .waitingForBrowser のまま維持）
            var pollInterval = TimeInterval(interval)

            while true {
                try await Task.sleep(nanoseconds: UInt64(pollInterval * 1_000_000_000))

                do {
                    let tokenInput = CreateTokenInput(
                        clientId: clientId,
                        clientSecret: clientSecret,
                        deviceCode: deviceCode,
                        grantType: "urn:ietf:params:oauth:grant-type:device_code"
                    )
                    let tokenOutput = try await oidcClient.createToken(input: tokenInput)

                    // 認証成功
                    guard let token = tokenOutput.accessToken else {
                        throw SSOAuthError.createTokenFailed
                    }

                    accessToken = token
                    let expiresIn = tokenOutput.expiresIn
                    if expiresIn > 0 {
                        accessTokenExpiry = Date().addingTimeInterval(TimeInterval(expiresIn))
                    }

                    loginState = .selectingAccount
                    return

                } catch let error as AuthorizationPendingException {
                    // authorization_pending: ポーリング継続
                    _ = error
                    continue
                } catch let error as SlowDownException {
                    // slow_down: ポーリング間隔を5秒延長
                    _ = error
                    pollInterval += 5
                    continue
                } catch let error as ExpiredTokenException {
                    // expired_token: タイムアウト
                    _ = error
                    loginState = .error("認証がタイムアウトしました。再度 SSO ログインを実行してください")
                    return
                }
            }

        } catch let error as SSOAuthError {
            loginState = .error(error.localizedDescription)
        } catch {
            loginState = .error("SSO 接続に失敗しました: \(error.localizedDescription)")
        }
    }


    // MARK: - アカウント一覧取得

    /// SSO で利用可能なアカウント一覧を取得する
    func fetchAccounts() async throws {
        guard let token = accessToken, isTokenValid else {
            loginState = .error("SSO セッションが期限切れです。再度 SSO ログインを実行してください")
            return
        }
        guard let region = ssoRegion else {
            loginState = .error("SSO リージョンが設定されていません")
            return
        }

        do {
            let ssoConfig = try await SSOClient.SSOClientConfiguration(region: region)
            let ssoClient = SSOClient(config: ssoConfig)

            let input = ListAccountsInput(accessToken: token)
            let output = try await ssoClient.listAccounts(input: input)

            let accountList = output.accountList ?? []
            accounts = accountList.compactMap { account in
                guard let accountId = account.accountId,
                      let accountName = account.accountName else { return nil }
                return SSOAccountInfo(accountId: accountId, accountName: accountName)
            }

            if accounts.isEmpty {
                loginState = .error("利用可能なアカウントがありません。IAM Identity Center の権限設定を確認してください")
            } else {
                loginState = .selectingAccount
            }
        } catch {
            loginState = .error("アカウント一覧の取得に失敗しました: \(error.localizedDescription)")
        }
    }

    // MARK: - ロール一覧取得

    /// 指定アカウントで利用可能なロール一覧を取得する
    /// - Parameter accountId: AWS アカウント ID
    func fetchRoles(accountId: String) async throws {
        guard let token = accessToken, isTokenValid else {
            loginState = .error("SSO セッションが期限切れです。再度 SSO ログインを実行してください")
            return
        }
        guard let region = ssoRegion else {
            loginState = .error("SSO リージョンが設定されていません")
            return
        }

        do {
            let ssoConfig = try await SSOClient.SSOClientConfiguration(region: region)
            let ssoClient = SSOClient(config: ssoConfig)

            let input = ListAccountRolesInput(accessToken: token, accountId: accountId)
            let output = try await ssoClient.listAccountRoles(input: input)

            let roleList = output.roleList ?? []
            roles = roleList.compactMap { $0.roleName }

            if roles.isEmpty {
                loginState = .error("このアカウントで利用可能なロールがありません")
            } else {
                loginState = .selectingRole
            }
        } catch {
            loginState = .error("ロール一覧の取得に失敗しました: \(error.localizedDescription)")
        }
    }

    // MARK: - 一時認証情報取得

    /// 指定アカウント・ロールの一時認証情報を取得する
    /// - Parameters:
    ///   - accountId: AWS アカウント ID
    ///   - roleName: ロール名
    func fetchCredentials(accountId: String, roleName: String) async throws {
        guard let token = accessToken, isTokenValid else {
            loginState = .error("SSO セッションが期限切れです。再度 SSO ログインを実行してください")
            return
        }
        guard let region = ssoRegion else {
            loginState = .error("SSO リージョンが設定されていません")
            return
        }

        do {
            let ssoConfig = try await SSOClient.SSOClientConfiguration(region: region)
            let ssoClient = SSOClient(config: ssoConfig)

            let input = GetRoleCredentialsInput(
                accessToken: token,
                accountId: accountId,
                roleName: roleName
            )
            let output = try await ssoClient.getRoleCredentials(input: input)

            guard let creds = output.roleCredentials,
                  let accessKeyId = creds.accessKeyId,
                  let secretAccessKey = creds.secretAccessKey,
                  let sessionToken = creds.sessionToken else {
                throw SSOAuthError.getRoleCredentialsFailed
            }

            // expiration はミリ秒単位の Unix タイムスタンプ
            let expiration: Date
            if creds.expiration != 0 {
                expiration = Date(timeIntervalSince1970: TimeInterval(creds.expiration) / 1000.0)
            } else {
                // デフォルト: 1時間後
                expiration = Date().addingTimeInterval(3600)
            }

            temporaryCredentials = SSOTemporaryCredentials(
                accessKeyId: accessKeyId,
                secretAccessKey: secretAccessKey,
                sessionToken: sessionToken,
                expiration: expiration
            )

            // キャッシュをファイルに保存
            saveCacheToFile()

            loginState = .authenticated
        } catch let error as SSOAuthError {
            loginState = .error(error.localizedDescription)
        } catch {
            loginState = .error("一時認証情報の取得に失敗しました: \(error.localizedDescription)")
        }
    }

    // MARK: - リセット

    /// ログイン状態を完全にリセットする
    func reset() {
        loginState = .idle
        accounts = []
        roles = []
        temporaryCredentials = nil
        accessToken = nil
        accessTokenExpiry = nil
        ssoRegion = nil
        clearCache()
    }
}

// MARK: - SSOAuthError

/// SSO 認証フローのエラー
enum SSOAuthError: LocalizedError {
    case registerClientFailed
    case startDeviceAuthorizationFailed
    case createTokenFailed
    case getRoleCredentialsFailed

    var errorDescription: String? {
        switch self {
        case .registerClientFailed:
            return "SSO 接続に失敗しました。Start URL と SSO リージョンを確認してください"
        case .startDeviceAuthorizationFailed:
            return "デバイス認証の開始に失敗しました。Start URL を確認してください"
        case .createTokenFailed:
            return "認証に失敗しました。再度 SSO ログインを実行してください"
        case .getRoleCredentialsFailed:
            return "一時認証情報の取得に失敗しました"
        }
    }
}
